#59 Additional validation for audio files.

benridge opened the pull request:
Moved deck and card validation into service. Added Tests.

Two new validations:
- verify the files referenced in the spreadsheet exist in the audio file. This is marked as a fatal error.
- verify if an audio file is used twice. Non-fatal.

I moved the verify() methods out of Deck, Language, and Cards to keep them as dumb as possible. Put the verify logic into VerifyDeckService and VerifyCardService. Language.verify() was just calling the card.verify() so that's done in VerifyDeckService now. 
That refactor should allow future verifications without bloating the model classes.

Please comment! No comment is too small.

benridge commented on src/test/java/org/mercycorps/translationcards/txcmaker/task/VerifyDeckTaskTest.java:
```diff
@@ -161,4 +164,11 @@ public void shouldDownloadAudioFiles() throws Exception {
+    public void shouldVerifyUsingService() throws Exception {
+        verifyDeckTask.verifyDeck(request);
+
+        verify(verifyDeckService).verify(drive, deck, AUDIO_DIR_ID);
```
yo dawg i heard you like verifying. So I verified that you are using the verify service to verify.

austin-garrard commented:
I really like the refactoring of the deck and card validation! It makes it much easier to see how the validation is being done. I can also see a clear path to adding more validation logic.

One quirk I noticed is that for duplicate audio file errors, only cards after the first one will have the error. My gut expects that error to show up on all the cards with that audio file. I think it's fine for now but I'm going to add it to the list of validation things to hash out. 

Another thing related to the duplicate error is that, as far as I know, two cards having the same audio file will crash the mobile app. So then the discussion becomes do we fix it in the app and leave this as non-fatal, or do we make this a fatal error because of existing behavior in the app. 

Also note that the UI currently doesn't let you publish a deck with non-fatal errors. 

All that said I'm fine with merging this, as long as we proceed to nail down some concepts/language about validation

benridge commented:
That's a good point about the duplicate files. If two cards use the same file it only marks one of the cards as having an issue, not both. I was trying to reduce the number of times the cards are looped through for validation but should have considered that. If it's not a showstopper we can revisit it later. Else I can work on it when I get a chance.

benridge commented:
I made a second commit based on @austin-garrard 's suggestions.
- Cards that have the same audio file will now show an error for all files, not just after the first one.
This required moving the validation to VerifyDeckService, which makes more sense.
- Duplicate audio error is considered a fatal error (so it now shows as red text).

Other changes
- Added a test to VerifyDeckTask that I had missed.
- Added and fixed up tests for VerifyDeckService and VerifyCardService.
- Added a constructor to Card so that errors can be initialized with an empty list (following the same pattern that Deck does).

As always, feel free to comment on any issues or unclear code, no matter how big or small.

PatrickDale commented:
@benridge I documented this in the Android application repo, but not here. I found that the errors we were getting were not from duplicate audio file names, but the way we were initializing the list of translations after switching languages in decks that have a different number of translations per language. I'm really sorry for the confusion here, and will make sure to update both boards in the future...

You can find more info on it here: https://github.com/translation-cards/translation-cards/issues/291

@austin-garrard @benridge What are your thoughts on keeping this validation as a warning rather than error? From the perspective of a deck maker, it could be nice to be notified if I have used the same audio file in multiple locations...

PatrickDale commented:
@benridge we noticed in your original PR that uploading the same audio file twice produces a "non-fatal" error. what did you mean by that? i ask because we checked out your branch and tried running that scenario, but we were unable to publish. thanks!

- @davsucks @patrickdale

benridge commented:
@PatrickDale thanks for looking at this. 
I can easily change the duplicate audio file to be a warning instead of an error. But right now warnings still block from publishing. The only difference between warning vs. error is the error message is yellow instead of red.  A bigger change (another user story IMHO) would be to make warnings non-fatal. So I shouldn't have said non-fatal, I meant a warning (which is still fatal i.e. unpublishable). We don't have UI design around warnings on the Preview page as far as I know. I can ask Olly for what that would look like, or if you have something pass it along and I can work on that. I suggest that be another follow-up story.

PatrickDale commented:
@benridge That sounds great, I will write up a story card for that. Thanks!
